### PR TITLE
[eclipse/xtext#1777] ported xtend code to java

### PR DESCRIPTION
**Pull request by cdietrich**
[eclipse/xtext#1777] ported xtend code to java
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+public class HiddenLeafAccess {
+	@Inject
+	private NodeModelAccess nodeModelAccess;
+	private int offset;
+
+	public HiddenLeafs getHiddenLeafsBefore(INode node) {
+		ILeafNode start = nodeModelAccess.findNextLeaf(node, l -> !l.isHidden());
+		List<ILeafNode> nodes = this.findPreviousHiddenLeafs(start);
+		if (start != null) {
+			int offset = 0;
```
Comment (tivervac) [inline org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:41]: You can put this two lines higher and remove line 49 as well

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+	}
+
+	protected HiddenLeafs newHiddenLeafs(int offset, List<ILeafNode> nodes) {
+		HiddenLeafs result = new HiddenLeafs(offset);
+		boolean trailing = true;
+		for (ILeafNode node : nodes) {
+			boolean comment = false;
+			int newLines = 0;
+			if (!Strings.isNullOrEmpty(node.getText())) {
+				for (Integer i : new IntegerRange(0, node.getText().length() - 1)) {
```
Comment (tivervac) [inline org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:64]: There are lot of `@Pure` in the file that declares `IntegerRange`

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+					char c = node.getText().charAt((i).intValue());
```
Comment (tivervac) [inline org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:65]: No need for the `intValue()` AFAIK

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+					if ('\n' == c) {
+						newLines = newLines + 1;
```
Comment (tivervac) [inline org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:67]: `newLines++` or `+=1`?

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+					} else {
+						if (!Character.isWhitespace(c)) {
```
Comment (tivervac) [inline org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:69]: You can just make this one `else if`

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+import com.google.inject.Inject;
+
+/**
+ * @deprecated use {@link ITextRegionAccess}
+ */
+@Deprecated
+public class HiddenLeafAccess {
+	@Inject
+	private NodeModelAccess nodeModelAccess;
+	private int offset;
```
Comment (tivervac) [inline org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:35]: Why is this a field? It's only used in one method

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+		}
+		return result;
+	}
+
+	protected ILeafNode findPreviousLeaf(INode node, Function1<? super ILeafNode, ? extends Boolean> matches) {
+		INode current = node;
+		while (current instanceof ICompositeNode) {
+			current = ((ICompositeNode) current).getLastChild();
+		}
+		if (current instanceof ILeafNode && matches.apply((ILeafNode) current).booleanValue()) {
```
Comment (tivervac) [inline org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:133]: No need for booleanValue()

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+			return (ILeafNode) current;
+		}
+		if (current != null) {
+			NodeIterator ni = new NodeIterator(current);
+			while (ni.hasPrevious()) {
+				INode previous = ni.previous();
+				if (previous instanceof ILeafNode && matches.apply((ILeafNode) previous).booleanValue()) {
```
Comment (tivervac) [inline org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:140]: No need for booleanValue()

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+		}
+		List<ILeafNode> result = new ArrayList<>();
+		if (current != null) {
+			NodeIterator ni = new NodeIterator(current);
+			while (ni.hasPrevious()) {
+				INode previous = ni.previous();
+				if (!Objects.equal(previous, current) && previous instanceof ILeafNode) {
+					if (((ILeafNode) previous).isHidden()) {
+						result.add((ILeafNode) previous);
+					} else { // if(!result.empty)
```
Comment (tivervac) [inline org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:161]: Seems like a pretty useless comment to me

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java (hunk 6 above)
Comment (cdietrich) [reply org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:35]: refactoring error

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java (hunk 9 above)
Comment (cdietrich) [reply org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:161]: did not want to loose it

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java (hunk 4 above)
Comment (cdietrich) [reply org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java:64]: didnt wanna change this

File: org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java (hunk 2 above)
Comment (tivervac) [reply org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java]: I guess you'll handle this in another MR? Don't see any changes now

Comment (cdietrich) [reply org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java]: ? replaced it with a for loop ?!?

Comment (tivervac) [reply org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java]: That also works :+1: 

What I meant by my comment is that the file that declares IntegerRange has already undergone the xtend->java process, but we forgot to remove the `@Pure annotations` so we should also get rid of those

Comment (cdietrich) [reply org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java]: no integerrange always was java. and its in xbase lib. thus keep the pure for xtend (only) users

Comment (tivervac) [reply org.eclipse.xtext.xbase/deprecated/org/eclipse/xtext/xbase/formatting/HiddenLeafAccess.java]: I see, makes sense

**Pull request merged**